processors/blockprocessor/block_processor.go

```diff
@@ -0,0 +1,367 @@
+	}
+
+	l, err := util.MakeLedger(proc.logger, false, genesis, pCfg.IndexerDatadir)
+	if err != nil {
+		return fmt.Errorf("could not make ledger: %w", err)
+	}
+	if l == nil {
+		return fmt.Errorf("ledger was created with nil pointer")
+	}
+
+	nextRound := initProvider.NextDBRound()
+	if uint64(l.Latest()) > uint64(nextRound) {
+		return fmt.Errorf("the ledger cache is ahead of the required round (%d > %d) and must be re-initialized", l.Latest(), nextRound)
+	}
+
+	if l.Latest() == 0 {
+		blk, cert, err := l.BlockCert(0)
+		if err != nil {
+			return fmt.Errorf("error getting block and certificate: %w", err)
+		}
+		vb := ledgercore.MakeValidatedBlock(blk, ledgercore.StateDelta{})
+
+		err = proc.addValidatedBlockCertToLedger(vb, cert, blk.Round())
+		if err != nil {
+			return fmt.Errorf("error adding gensis block: %w", err)
+		}
+	}
+
+	proc.ledger = l
+	proc.ctx = ctx
+	proc.cfg = cfg
+
+	return nil
+}
+
+func (proc *blockProcessor) extractValidatedBlockAndPayset(blockCert *rpcs.EncodedBlockCert) (ledgercore.ValidatedBlock, transactions.Payset, error) {
+
+	if blockCert == nil {
+		return ledgercore.ValidatedBlock{}, transactions.Payset{}, fmt.Errorf("cannot process a nil block")
+	}
+	if blockCert.Block.Round() != (proc.ledger.Latest() + 1) {
+		return ledgercore.ValidatedBlock{}, transactions.Payset{}, fmt.Errorf("invalid round blockCert.Block.Round(): %d nextRoundToProcess: %d", blockCert.Block.Round(), uint64(proc.ledger.Latest())+1)
+	}
+
+	// Make sure "AssetCloseAmount" is enabled. If it isn't, override the
+	// protocol and update the blocks to include transactions with modified
+	// apply data.
+	proto, ok := config.Consensus[blockCert.Block.BlockHeader.CurrentProtocol]
+	if !ok {
+		return ledgercore.ValidatedBlock{}, transactions.Payset{}, fmt.Errorf(
+			"cannot find proto version %s", blockCert.Block.BlockHeader.CurrentProtocol)
+	}
+	protoChanged := !proto.EnableAssetCloseAmount
+	proto.EnableAssetCloseAmount = true
+
+	ledgerForEval := indexerledger.MakeLedgerForEvaluator(proc.ledger)
+
+	resources, err := prepareEvalResources(&ledgerForEval, &blockCert.Block)
+	if err != nil {
+		proc.logger.Panicf("ProcessBlockCert() resources err: %v", err)
+	}
+
+	delta, payset, err := ledger.EvalForIndexer(ledgerForEval, &blockCert.Block, proto, resources)
+
+	if err != nil {
+		return ledgercore.ValidatedBlock{}, transactions.Payset{}, fmt.Errorf("eval err: %w", err)
+	}
+
+	// validated block
+	var vb ledgercore.ValidatedBlock
+	if protoChanged {
+		block := bookkeeping.Block{
+			BlockHeader: blockCert.Block.BlockHeader,
+			Payset:      payset,
+		}
+		vb = ledgercore.MakeValidatedBlock(block, delta)
+		return vb, payset, nil
+	}
+	vb = ledgercore.MakeValidatedBlock(blockCert.Block, delta)
+	return vb, blockCert.Block.Payset, nil
+}
+
+func (proc *blockProcessor) addValidatedBlockCertToLedger(vb ledgercore.ValidatedBlock, cert agreement.Certificate, roundToWait basics.Round) error {
```

Comment:
nit: merge with `OnComplete`, seems like it no longer needs to be a separate function.

Reply:
I think it does since it is used in `OnComplete` and in `Init`

Reply:
`Init` would call `OnComplete` instead